### ISSUE TITLE
content is read as octets, not characters, with no concept of decoding

### ISSUE THREAD

**Issue by karenetheridge** [Perl-Critic/PPI#22]
`new($filename)` reads the file as bytes, with no encoding layers, so any content that isn't Latin1 will cause issues.  `new(\$content)` is read exactly the same way.

Files need to be decoded, so an encoding parameter is needed. Content strings either need to be decoded similarly, or be passed as already-decoded characters.

I would suggest new APIs that don't conflict with the existing ones, to try to preserve backcompat as much as possible.

**Comment by 0x62ash** [comment]
Are there any progress?

**Comment by karenetheridge** [comment]
The right thing to do here is what perl itself does -- when `use utf8` is seen in code, re-parse the entire document from the beginning treating all octets as utf8-encoded.

(See also #168, #26.)

**Comment by wchristian** [comment]
@0x62ash Not really, because i'm stuck on other things and $work and health. Basically, looking for round tuits here.

@karenetheridge Sadly Perl does not act like that. utf8.pm is lexically scoped and only affects whatever comes after it and is not in a higher scope.

And that is what makes implementing a useful solution for this a bit of a rat's nest.

**Comment by karenetheridge** [comment]
For filenames, we'll have to pass the right binmode to use, e.g. `my $ppi = PPI::Document->new("Foo/Bar.pm", encoding => ':encoding(UTF-8)');`

It's also possible (although maybe not desirable from a design perspective) to also support passing a filehandle, so the user can specify his own encoding on the fh passed in.  You'd need to do a 'if blessed, and quacks like a IO::something' check and I'd ask someone like @leont what the right semantics are for that (if not smelly).

Also,

**Comment by karenetheridge** [comment]
At the very least, the PPI::Document documentation needs to make it clear (for the reference-to-string interface) whether it wants a string of decoded characters or a string of encoded bytes.  I suspect it is the former (since it does stringy operations throughout, e.g. regular expressions, substr etc), but it should be clear.  Then the `->new($filename)` interface can be fixed to handle an encoding parameter.

**Comment by rjbs** [comment]
@karenetheridge Perl does **not** rescan from the beginning when `use utf8` is encountered.  Behold:

```perl
use 5.12.0;
my $x = "Queensrÿche";
use utf8;
my $y = "Queensrÿche";

say length $x;
say length $y;
```

This prints 12, 11.  utf8 is lexical in scope, so your file can switch between Latin-1 and UTF-8 at will.  Yup.

**Comment by rjbs** [comment]
We (Christian, Karen, and I) discussed this problem at the 2016 QAH in Rugby.  One thing that came up was the BOM.  The way it ties in is this:

When `perl` sees a file starting with a [BOM](https://en.wikipedia.org/wiki/Byte_order_mark), it switches to the right mode of parsing: UTF-16BE or LE.  If it sees a UTF-8 BOM, it ignores it.  That is: it does *not* behave like it saw "use utf8".  The UTF-8 BOM is an anomaly, and its use is discouraged in general.

For example:

> The IETF recommends that if a protocol either (a) always uses UTF-8, or (b) has some other way to indicate what encoding is being used, then it "SHOULD forbid use of U+FEFF as a signature."

The problem is that when reading a file as octets, as PPI does when reading by filename, the UTF-8 encoded BOM becomes a PPI::Token::BOM and is harmless.  Unfortunately, all kinds of other things are broken, like `$π = 3`.  So, the solution may be to slurp the file, decode it, and pass the decoded string to `PPI::Document->new(\$str)`.  This will fix many things, but the BOM, now a U+FEFF, will cause a failure to parse.  It becomes the responsibility of the file-reader to deal with the BOM.

Probably this is all unavoidable.  The BOM in a UTF-8 file doesn't cause Perl to turn on UTF-8 tokenization, so it isn't sensible that it would do so in PPI.  `utf8.pm` *does* turn on UTF-8 tokenization, but I'm not sold on the idea of PPI tracking the effect of a lexical pragma.  I guess it would be useful, though.

PPI can't know whether a string reference indicates that it's from a file at offset zero, so it's not going to be clear to PPI whether a leading U+FEFF is a BOM or just whitespace.

**Comment by karenetheridge** [comment]
I'd be quite happy to see "switching encodings in the middle of a document is not supported". It would even be fine for the caller to specify what encoding to read a given filename with, so we don't have to guess (and ignoring BOMs).  A loud warning if `use utf8` is seen, and we're parsing raw/latin1, would also be ok as a compromise.

The biggest problem, IMO, is that `->new($filename)` non-negotiably parses in latin1, so if a different encoding is being used the user has to slurp the file themselves into a string with the correct encoding and then pass `->new($content_string)`.

**Comment by xsawyerx** [comment]
I'm just hitting this now with the following case:

```
use utf8;
use PPI::Document;

my $doc = PPI::Document->new(\"± values");
# $doc = undef
```

Is there a recommendation on how to handle this at the moment until PPI resolves this internally?

**Comment by karenetheridge** [comment]
Pass the file content through `utf8::encode(..)` first before sending it to PPI::Document -- but depending on the position of the wide character(s), this may no longer be valid perl.  If it's just comments or literal strings you'll be ok, but if the wide chars are in identifiers then the regex for that token type will no longer match.  (And then if you're doing something with the bits you parse out of the document with PPI, utf8::decode it again.)

e.g.:

    # wide char ± in comment - ok
    my $str = "wide char ± in string - ok!";
    my $±; # wide char in token: PPI will not like this

**Comment by SineSwiper** [comment]
Oh, great.  I can't release a module to our DarkPAN because:

    Fatal error... regex failed to match in 'É�uoâ��doÉ¥É�              ( $  ) ;
    ' when expected at /usr/local/gsgperl/dev/active/lib/perl5/PPI/Token/Word.pm line 240.
     while processing file lib/SOME/MODULE/Value_Maps.pm at 
    /usr/local/gsgperl/dev/active/lib/perl5/Dist/Zilla/Plugin/MetaProvides/Package.pm line 199.

Why can't PPI just act like real Perl and re-read the file when it encounters `use utf8;`?

**Comment by karenetheridge** [comment]
@SineSwiper are you using the latest version of `[MetaProvides::Package]`? line 199 doesn't seem to call into PPI, in the version I have... and a year or two ago at a toolchain hackathon I patched Dist::Zilla::Role::PPI so it would more-or-less work with the current state of PPI - it will parse utf8-encoded tokens, but most of the time this is ok (and should be fine for your purposes, as your `package` and `$VERSION` declarations will be pure ascii).  Please upgrade your various dzil modules and see if you see different behaviour?

**Comment by SineSwiper** [comment]
Thanks for the speedy reply, Karen.  Yes, an upgrade fixed that bit.

For the record, I think P5P is overthinking this, especially with the Queensrÿche example.  Just solve for the `use utf8` case, which is going to be 99.9999% of it, and worry about the edge cases in other tickets.

After all, "only Perl can parse Perl".  PPI is supposed to be the "good enough" parsing module.  Perfect is the enemy of good, etc.

**Comment by adamkennedy** [comment]
Because it was bloody hard to read Perl at all, and I took a crapton of
shortcuts to make the parsing speed not quite so bad as it would have been,
like a bunch of asc()ord() lookups instead of handing every character in
Unicode.

We also refuse to execute code, and fundamentally, "use" statements run
code.

Finally, by the time I got to proper Unicode handling as a priority I was
so burned out I didn't have the heart for that deep a change.

Patches probably kinda sorta welcome...

**Comment by xsawyerx** [comment]
@adamkennedy Good to hear from you. :)

I wrote a harness script that runs Perl::Critic in parallel and I've just added a simple: `$critic->critique( \Path::Tiny::path($filename)->slurp_utf8 )` and it works perfectly. The only problem is the speed.

I think speed could be improved by taking particular hot code-paths (and repeated behavior) in Perl::Critic and implementing it in XS.
